### DISCUSSION TITLE
class pointer sugar

### DISCUSSION THREAD

**Discussion by HKhademian** [carbon-language/carbon-lang#1545] [Language design]
if we have a class, like: `class A { var x: i64; }`
and a variable from it, like: `let a: A = ...;`.
then we can access `x` via dot, like: `a.x = 5;`;

if we make a pointer from that, like: `let b: A* = &a;`
there is two way to access `x`,
first via dereference+dot, like: `(*b).x`
second via syntax sugar, like: `b->x`

What is the downside of implicit dereferencing in this scope? like: `b.x`
Why can't the compiler automatically detect `<ptrOfX>.<fieldOfX>` and treat it like a normal `<objOfX>.<fieldOfX>`?

Is there any benefits in explicit dereferencing ?
There is a good [youtube video](https://www.youtube.com/watch?v=yKI-VOBBFu8) from [Tsoding](https://github.com/tsoding), showing how easy it is to add this feature to TinyC, and how convenience it is to program without `->` torture.

**Comment by HKhademian** [comment]
here is a simple example (inspired by that video):

You start programming for your new side project to forget about 100 previous.
it's a simple game and you want to keep track of a mine field:
```carbon
class MineField {
  var width: u64;
  var height: u64;
}
```

then you create a function to calculate area of that field:
```carbon
fn calculateMineArea(mf: MineField) -> u64 {
  // imaging there is many of mf.X usage
  return mf.width * mf.height;
}
```

after long programming session, you see you need a huge additional data to keep in `MineField`, so:
```carboon
class MineField {
  var width: u64;
  var height: u64;
  var newLargeDate1: u256;
  var newLargeDate2: u256;
  var newLargeDate3: u256;
  var newLargeDate4: u256;
}
```

So it's obvious to you that, now it's inefficient to copy whole object to just calculate the area, you decide to use pointer instead:
```carbon
fn calculateMineArea(mf: MineField*) -> u64 {
  // ...
}
```

Now you realize that you need to replace all those `<obj> . <field>` with `<ptr> -> <field>`.

Another similar situation is you need to convert a pointer-arg to a value-arg (for example for better performance - no pointers , ...), now you need to convert all `<ptr> -> <field>` to `<obj> . <field>`

Imagine in a large code bases, with many functions to convert, how it's inefficient this process.
and how many lines of code that need to changed (reviewed, examined, approved), just to simply convert a normal value argument to pointer one.

**Reply by YagaoDirac** [reply]
Emmm, you perfectly explained how to use -> in c++ but I guess this discussion is not about how to use -> in c++. It's about can we replace the -> with . in some/all cases. 
I also commented below. I think it's practical in some cases.

**Comment by jwtowner** [comment]
I'm thinking I agree with this. Hasn't the general trend in C++ been towards value types and hiding pointers (and smart pointers) as implementation details, especially in Boost and the Standard library? The concept-model idiom has been around for over a decade and more recently the C++ [metaclasses](https://wg21.link/p0707) proposal was introduced to allow for the autogeneration of polymorphic type-erased facades, with pointer semantics and lifetime details abstracted away (similar to Rust's `dyn` keyword).

Maybe I misunderstand, but the trajectory that C++ is taking seems to be odds with Carbon's choice of syntax here. From reading the available documentation and proposals for Carbon, it looks like there are plans to enable type erasure, metaprogramming and generation of type-erased facets from archetypes. But is there still a way that Carbon can allow for implicit pointer semantics with `.` syntax, as is done in C++ with l-value references? If not, how are things going to look when interoperating with C++ components that use the concept-model idiom. Using `->` syntax almost seems like step backwards for Carbon, although I definitely understand the desire to try to simplify the choice of references vs pointers.

**Reply by YagaoDirac** [reply]
I think the trend is not hiding pointers. A lot design is to provide more flexibility and allow less code and safer code.
Pointers work differently in different case. Some are basically another way to implement the reference.
Only when pointers are used to handle job inside containers and heavy low-level calculation, the arithmetic of pointers is meaningful. And this is also the dangerous case. 
When the containers are implemented well enough, people only need to manipulate them from outside, only the ref-like pointers are needed. This is probably the reason you find more pointers are hidden. 
As long as Carbon needs high performance and heavily customized containers, pointer is probably the only choice.

**Comment by geoffromer** [comment]
I agree that there's a real problem here, but I'm not sure if this sort of auto-dereferencing is a viable way of solving it. The problem is that Carbon, like C++, will probably have a lot of smart pointer types like `unique_ptr`, and in fact we even expect the "plain" pointer type to actually be a class type (because [all types will be library types](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/library_apis_only.md)). So if we allow auto-dereferencing, then a syntax like `p.foo` has several possible meanings: it could mean `p.foo`, or `(*p).foo`, or `(**p).foo`, etc.

Of course, we could come up with rules for disambiguating among those expressions, but there would still be a risk that the meaning the author intended (or the meaning the reader understands) is different from the meaning that the disambiguation rules choose, leading to unpleasant surprises (anecdotally, I've heard that this can happen in Rust). It also creates problems for [software evolution](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#software-and-language-evolution), because it means that adding a method `foo` to a pointer type always risks breaking client code, even at run time, by changing the meaning of `p.foo` in client code.

**Reply by staniek** [reply]
> So if we allow auto-dereferencing, then a syntax like p.foo has several possible meanings: it could mean p.foo, or (*p).foo, or (**p).foo, etc.

Yes, and I think one of the blessings of the languages like C++ is that they raise compiler error in places during refactoring and general programming practice (so also for the missing changes from "." to "->").

For me there are not many worse moments than compiler (or linker) silently **accepting** a flawed source code change without raising an error but "delaying" the actual process of debugging to the run time.

**Reply by HKhademian** [reply]
In any context , your `p` variable is `Ball p` , `Ball *p` , `Ball **p` or ....
So the auto-dereferencing for a reference never cause an ambiguity between `*p` or `p` access.
Also please look at how references in c++ provide dot access. so Why not pointers do?

Also you mentioned that `Ball **p`,
Can `->` provide access for that ?!! NO
In all documents, `p->x` operator defines as syntax sugar for `(*p).x`,
so If you apply it to `Ball **p` like `p->x` then it translated to `(*p).x` which `*p` type is `Ball *` (a pointer), and it's illegal to access  that with a dot!

I just said why `p->x` when compilers are smart enough to automatic convert `ptr.x` to `(*ptr).x`

**Reply by HKhademian** [reply]
@staniek why it's ok for you to use the same syntax for `Ball p` and `Ball &p`?
What if somebody changes your function argument from `vector &v` to `vector v`?
And if all of them are ok to you, then why do you think it's bad practice for pointers.

**Reply by staniek** [reply]
>  why it's ok for you to use the same syntax for `Ball p` and `Ball &p`?

It is not OK to me, it's C++'s similar flaw. That's why I'd be unhappy if we copy that to the new language.

> What if somebody changes your function argument from `vector &v` to `vector v`?

I've been there, and failed many times, eg. flaws slipped through the reviews. This makes consumption of the language harder.

**Reply by OlaFosheimGrostad** [reply]
How about  #1736 ?

**Reply by staniek** [reply]
> How about #1736 ?

No opinion for now. I am coming from the Qt universe and I have not reviewed the implications for existing code wrt clarity and confusion.

**Reply by geoffromer** [reply]
> In any context , your `p` variable is `Ball p` , `Ball *p` , `Ball **p` or ....
> So the auto-dereferencing for a reference never cause an ambiguity between `*p` or `p` access.

Here's an example in C++:

```c++
struct Ball {
  void get() const;
};

auto my_ptr = std::make_unique<Ball>();
```

`my_ptr.get()` and `(*my_ptr).get()` are both valid, but call different methods on different objects, so if `my_ptr.get()` could also mean `(*my_ptr).get()`, it would be ambiguous.

> Also please look at how references in c++ provide dot access. so Why not pointers do?

The reason C++ references can do this is that they cannot have methods: operations on a reference are always operations on the object it refers to, and never operations on the reference itself.

> Also you mentioned that `Ball **p`,
> Can `->` provide access for that ?!! NO
> In all documents, `p->x` operator defines as syntax sugar for `(*p).x`,
> so If you apply it to `Ball **p` like `p->x` then it translated to `(*p).x` which `*p` type is `Ball *` (a pointer), and it's illegal to access that with a dot!

Sorry, I assumed you were proposing that auto-dereferencing can be applied any number of times, because my understanding is that's how it works in Rust.

**Reply by OlaFosheimGrostad** [reply]
> Here's an example in C++:
> 
> ```c++
> struct Ball {
>   void get() const;
> };
> 
> auto my_ptr = std::make_unique<Ball>();
> ```
> 
> `my_ptr.get()` and `(*my_ptr).get()` are both valid, but call different methods on different objects, so if `my_ptr.get()` could also mean `(*my_ptr).get()`, it would be ambiguous.

If my suggestion #1736 was used then you could do it like this:

```
m_ptr.get();    // access Ball::get()
(&m_ptr).get(); // access std::unique_ptr<Ball>::get()
```

**Reply by L4stR1t3s** [reply]
> If my suggestion https://github.com/carbon-language/carbon-lang/issues/1736 was used then you could do it like this:
> ```
> m_ptr.get();    // access Ball::get()
> (&m_ptr).get(); // access std::unique_ptr<Ball>::get()
> ```

How does this work? How is the compiler supposed to know that it has to access Ball::get() when it encounters m_ptr.get()? unique_ptr is just a template right which results in a member Ball*. So how does the compiler know what member of the template class to pick? unique_ptr<> would have to have some sort of operator to define the behaviour. Or do you expect unique_ptr to be built into the language and get special treatment? I'd seriously advise against that because using a custom smart pointer implementation is quite a common practice.

It also feels very unintuitive that in order to access the class of a variable, you first have to &m_ptr it. How would a user know what m_ptr.get() does without knowing the internal workings of unique_ptr?

**Reply by OlaFosheimGrostad** [reply]
> How does this work? How is the compiler supposed to know that it has to access Ball::get() when it encounters m_ptr.get()? unique_ptr is just a template right which results in a member Ball*.

The same way you test traits. If `decltype(m_ptr)` is a builtin pointer or provides `operator->` or provides `operator*` then it is a "de-ref-able" expression. And the deref-type-constructor is applied. To get to the smart-pointer you have to undo the "deref" with an "underef". The deref-operator is `*`, and the underef-operator is `&`. Everything else follows from that.

**Reply by OlaFosheimGrostad** [reply]
Or you can think of it this way: in order to access the smart-pointer fields you take the address of the smart-pointer object and that will be auto-dereferenced, which gives you the smart-pointer object and not the pointed-to object.

**Reply by geoffromer** [reply]
> If my suggestion #1736 was used then you could do it like this:
> 
> ```
> m_ptr.get();    // access Ball::get()
> (&m_ptr).get(); // access std::unique_ptr<Ball>::get()
> ```

Sure, but that doesn't really change the fact that `m_ptr.get()` can mean different things depending on whether `m_ptr` and/or `*m_ptr` has a `get` method, and consequently:
- Readers might misunderstand what `m_ptr.get()` does, because they're unaware of one of those methods.
- `m_ptr.get()` might not do what the writer want, because they're unaware of one of those methods.
- `m_ptr.get()` might change meaning if a `get` method is added to one of those types.

And if we want to guarantee that those problems will be caught at compile time, we have to treat ambiguity in `m_ptr.get()` as a fatal error, and provide two additional syntaxes: one that unambiguously access a member of the pointer, and one that unambiguously accesses a member of the pointee.

**Reply by L4stR1t3s** [reply]
> in order to access the smart-pointer fields you take the address of the smart-pointer object and that will be auto-dereferenced

So the order of members becomes quite important then?
```
template <typename T> class shared_ptr
{
    T* ptr;
    size_t refcount;
}
```
```
template <typename T> class shared_ptr
{
    size_t refcount;
    T* ptr;
}
```

**Reply by OlaFosheimGrostad** [reply]
> Sure, but that doesn't really change the fact that `m_ptr.get()` can mean different things depending on whether `m_ptr` and/or `*m_ptr` has a `get` method, and consequently:

Not sure I understand. In Rust it first looks in the smart_pointer and then in the deref-result, so it is essentially a scope-within-a-scope lookup. That will give this kind of confusion. I absolutely do not recommend this approach.

In my issue #1736, you will get an error if there is no get() method:

1. `m_ptr.get()` will error if the pointed-to-object has no get() method and no lookup is done in the smart-pointer object.
2. `(&m_ptr).get()` will error if the smart-pointer has no get() method and no lookup is done in the pointed-to-object.

> * `m_ptr.get()` might change meaning if a `get` method is added to one of those types.

I don't see how as there is only one scope that is looked into.

> And if we want to guarantee that those problems will be caught at compile time, we have to treat ambiguity in `m_ptr.get()` as a fatal error, and provide two additional syntaxes: one that unambiguously access a member of the pointer, and one that unambiguously accesses a member of the pointee.

There is no ambiguity?

**Reply by L4stR1t3s** [reply]
> There is no ambiguity?
```
class Empty
{
    int get() { return 0; }
};
Empty e;
```
There is a member function get() in Empty and there is no member variable to point to.
If I want to access get() in Empty, do I have to write this?
`(&e).get()`
Or does this suffice?
`e.get();`
The first option is ridiculous, syntax wise. And the second option would mean that a developer would need to know exactly how every object he uses is constructed to know what object.xyz() would do.

**Reply by OlaFosheimGrostad** [reply]
> Or does this suffice?
> `e.get();`

This is sufficient since `Empty` does not provide a dereferencing mechanism. You only need the `(&obj).field` syntax if `obj` provides a dereferencing mechanism.

**Reply by L4stR1t3s** [reply]
And that's not ambiguous? How do you know if something has a dereferencing mechanism or not?

**Reply by OlaFosheimGrostad** [reply]
That is not ambiguous, that's right.

You know if something has a dereferencing mechanism in the same way you know that something is a smart pointer or an iterator.

**Reply by urielgarciarivas** [reply]
What about, instead of allowing auto-dereferencing for already implemented operators like `p.foo`, but creating a new operator that does just that? An example could be `p->>foo`.

**Reply by L4stR1t3s** [reply]
Oh, you just know?
```
Object obj;
obj.get();
```
Dereferencing mechanism or not?

**Reply by L4stR1t3s** [reply]
> What about, instead of allowing auto-dereferencing for already implemented operators like p.foo, but creating a new operator that does just that? An example could be p->>foo.

Are you for real?

**Reply by urielgarciarivas** [reply]
> Are you for real?

Yes, it could save a lot of dereferencing in C++ code-bases and the users would immediately notice it is an auto-dereference operator. The example operator I just gave was nothing more than the first thing that came into my mind, if you were wondering about the syntax there.

**Reply by L4stR1t3s** [reply]
You suggest adding ->> to dereference, in a discussion about removing -> to dereference?

**Reply by geoffromer** [reply]
@L4stR1t3s Your last few posts in this thread come across as sarcastic or mocking, which our [code of conduct](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md) doesn't permit (see especially "Be friendly and patient"). I've edited your last post to remove a particularly clear-cut instance.

**Reply by OlaFosheimGrostad** [reply]
> Oh, you just know?
> 
> ```
> Object obj;
> obj.get();
> ```
> 
> Dereferencing mechanism or not?

First of all, for your example to make sense you need to provide a use case where you would want to access the internals of a smart-pointer, iterator or similar without knowing what the type is. I haven't found such a case in the standard library or in any of my own code. There might be cases in generic code, which I am not aware of, but since we only talk Carbon code all such code would be new and you could guard against that using a trait.

So the remaining case would be instances where people have added dereferencing overloads to classes nilly-willy without having them being a main feature of the type. To what extent that is the case, it is still unlikely that it would not lead to an error. It is possible to find the probability of this not leading to an error by analyzing existing code bases on github.

When it comes to dereferencing of Carbon classes then you could provide a more robust mechanism for defining smart-pointers and avoiding the need for accessing the internals of the smart-pointer object through the `.` operator.

**Reply by L4stR1t3s** [reply]
> @L4stR1t3s Your last few posts in this thread come across as sarcastic or mocking, which our [code of conduct](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md) doesn't permit (see especially "Be friendly and patient"). I've edited your last post to remove a particularly clear-cut instance.

Come on, I'm supposed to pretend like suggesting ->> instead of -> isn't trolling?

**Reply by L4stR1t3s** [reply]
@OlaFosheimGrostad
```
Object obj;
obj.get();
```
Smart pointers? Iterators? Where do you see those in this code? You have suggested that the . operator does something different depending on arbitrary conditions. That's a bad idea, because as a developer you never know what the most basic operator will do without an investigation. There is a reason that the . operator can't be overridden in C++, and the -> was added to be overridable instead.

**Reply by OlaFosheimGrostad** [reply]
> Smart pointers? Iterators? Where do you see those in this code? You have suggested that the . operator does something different depending on arbitrary conditions.

There is nothing arbitrary going on. 

> That's a bad idea, because as a developer you never know what the most basic operator will do without an investigation.

If the programmer does not know what the operators do on the type, then s/he probably shouldn't use it?  It is true that abuse of operators in libraries could be an issue, but the only way to know what that means in practice is to analyze existing code, which to some extent can be done quantitatively.

It is in general difficult to improve on C++, but if you say "no change is possible", then you for certain will never find a design for a better system language. So even if this turns out to be too difficult for C++, you might still find a solution for auto-deref for Carbon defined classes.

Design isn't a linear exercise, it is a "circular" process.

**Reply by L4stR1t3s** [reply]
> If the programmer does not know what the operators do on the type, then s/he probably shouldn't use it?

That's the whole point, you always know what the . operator does in C++: access the members of the object. It never does anything else.
If you make the behaviour of the . operator dependent on arbitrary (yes, arbitrary because anyone can assign automated dereferencing to anything for whatever reason) properties than it becomes so much more complicated to write the simplest code.
`a.get();`
Does this access the class of a? Does this access something else? What if the something else itself has automated dereferencing? Where do I eventually wind up before get() is called? It's a nightmare to even consider having to program this way.

**Reply by OlaFosheimGrostad** [reply]
>  `a.get();` Does this access the class of a? Does this access something else? What if the something else itself has automated dereferencing? Where do I eventually wind up before get() is called? It's a nightmare to even consider having to program this way.

Automatic dereferencing only happens on the first level,  it will not happen if the object already has been dereferenced.

**Reply by L4stR1t3s** [reply]
Ok. Good luck with that.

**Reply by geoffromer** [reply]
@OlaFosheimGrostad Oh, OK, I think I understand why there's no ambiguity with your approach, but I'm also not very clear on what is gained with that approach, other than saving one keystroke in the common case.

**Reply by OlaFosheimGrostad** [reply]
@geoffromer  Well, there seems to be a demand for more coherent visual syntax for field access, and since the approach of Rust and D leads to names in the pointer shading names in the object I felt that I could come up with something a bit more principled. Anyway, it was an interesting exercise. ;)

If you want to do this only for Carbon, then I would instead suggest that you consider doing all interaction with the smart-pointer through free functions instead of members. E.g. `release(smart_pointer)` instead of `smart_pointer.release()`. *\*shrugs\**

**Comment by L4stR1t3s** [comment]
The original reasons for the existence of -> don't even exist anymore. But it has since morphed into something completely different. I think the most important difference between . and -> in C++ is the fact that the -> operator can be overriden and the . operator can not. Which brings us to smart pointers, that override the -> operator. This means that there is no syntactic difference between accessing a pointer and a smart pointer implementation:
```
X* x = new X;
x->func();
```
and
```
ptr<X> x(new X);
x->func();
```
This means that if you change from a regular pointer to a smart pointer or vice versa, no other code changes are required. Overriding the . operator sounds like it will be more trouble than its worth. You know what . does, and you also know that -> could be overridden. If only . operator exists and can be overridden, you can't trust anything you do with an object without checking if it overrides . operator. Not to mention the fallout if someone adds a . operator override in the future to an existing object.
This also brings us to a case where you need both -> and . operators in C++:
```
class X
{
    ...
   int get() { ... }
    ...
}
```
and
```
template <class T>
class ptr
{
    ...
    T* get() { ... };
    ...
}
```
Both the smart ptr template and X have a member function get. So does the following code access the get of X or ptr<X> if you only have the . operator?
```
ptr<X> x(new X);
auto y = x.get();
```
You could say that you need to dereference the object to make a distinction:
```
x.get(); // ptr<X>.get
(*x).get() // X.get()
```
But this leads to pretty ugly code in some case:
`(*(*x).y).z`
instead of
`x->y->z`
And then you can also never allow x.func() for accessing X.func(). Because if someone adds func() to ptr<> in the future you suddenly get completely different behaviour.
So I think there is a reason to have both -> and . operators.

**Comment by YagaoDirac** [comment]
Probably is because operator-> does concatenation while "." does plain offset.
//c++ code
class A{ int b;}
class C{
A* d; 
A* operator->(){return d;}
}
int main(){
A obj_A{};
C obj_C{};
obj_C.d = &obj_A;
obj_C->b; // the -> first finds the d and because d is also a pointer, the -> search inside it for the field of b.
If the example is wrong, ok, I'm not pretty sure about this. I don't write container libraries so, maybe I messed up.
But the dot operator works like:
obj_C.d->b; // dot first finds d and then d is a pointer, so we need a -> to find b.
But if we don't need the cascade, then -> is not needed.

For the cases with only 1 layer, dot and -> result the same. If you prefer dot than ->, it's possible to replace it in such case. But with more layers, -> is still a thing.

My personally prefer ".." rather than ->. Double dots and the classic single dot are similar and easier to type. Readability should not be an issue. The only trouble is that, when a->->b, double dots works like, a....b, probably a space is needed in the mid, like, a.. ..b. But since we can encourage people NOT to do this, just like very very rarely we see any ->-> in c++.

Short version:
-> does concatenation.
If there is only 1 layer, they are the same. Maybe we can replace -> with . in this case.
But when concatenation is needed, -> is still good enough.
c++:
a.b->c; // OK
if a.operator->() returns b, then
a->c; // OK, concatenation.
This example has 2 layers.
